// vue-docgen-api/dist/utils/getParseBabel.js
"use strict";var babel = require('@babel/core');
var presetEnv = require('@babel/preset-env');
var pluginObjectRestSpread = require('@babel/plugin-proposal-object-rest-spread');
var pluginTransformVueJsx = require('babel-plugin-transform-vue-jsx');
var path = require('path');
var process = require('process');

module.exports = function getJsxBabel(code, filename) {var comments = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
  // Provide filename and cwd to babel for:
  //  a) Proper loading of .babelrc
  //  b) Error messages saying where any SyntaxErrors are
  var cwd = process.cwd();
  var filenameRelative = path.relative(cwd, filename);

  var options = {
    ast: false,
    comments: comments,
    filename: filename,
    filenameRelative: filenameRelative,
    presets: [
    [
    presetEnv,
    {
      targets: {
        chrome: 52 } }]],




    plugins: [pluginObjectRestSpread, pluginTransformVueJsx],
    sourceRoot: cwd };


  return babel.transform(code, options);
};
