// vue-docgen-api/dist/utils/getSlots.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = getSlots;var _htmlparser = _interopRequireDefault(require("htmlparser2"));

var HtmlParser = _htmlparser.default.Parser;
var lex = require('pug-lexer');
var parse = require('pug-parser');
var wrap = require('pug-runtime/wrap');
var generateCode = require('pug-code-gen');

function getSlots(parts) {
  var output = {};
  if (parts.template && parts.template.content) {
    var template = parts.template.content;
    var lastComment = '';

    if (parts.template.attrs.lang === 'pug') {
      var funcStr = generateCode(parse(lex(parts.template.content)), {
        compileDebug: false,
        pretty: true,
        inlineRuntimeFunctions: false,
        templateName: '_parse' });

      var func = wrap(funcStr, '_parse');
      template = func();
    }

    var parser = new HtmlParser({
      oncomment: function oncomment(data) {
        if (data.search(/\@slot/) !== -1) {
          lastComment = data.replace('@slot', '').trim();
        }
      },
      ontext: function ontext(text) {
        if (text.trim()) {
          lastComment = '';
        }
      },
      onopentag: function onopentag(name, attrs) {
        if (name === 'slot') {
          var nameSlot = attrs.name || 'default';
          output[nameSlot] = {
            description: lastComment };


          lastComment = '';
        }
      } });


    parser.write(template);
    parser.end();
    return output;
  }
  return {};
}
